The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.

Version: 1.0.1

We are trying to update a Customer Custom type by removing a custom field. The following is the terraform plan:

```hcl
─────────────────

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # commercetools_type.regular-customer will be updated in-place
  ~ resource "commercetools_type" "regular-customer" {
        id                = "bfc50073-8989-4c96-8f4d-afb2cbfd93e0"
        name              = {
            "en-US" = "Regular Customer Type"
        }
        # (4 unchanged attributes hidden)

      - field {
          - input_hint = "SingleLine" -> null
          - label      = {
              - "en-US" = "A value-added tax identification number for value-added tax purposes."
            } -> null
          - name       = "vatNumber" -> null
          - required   = false -> null

          - type {
              - name   = "String" -> null
              - values = {} -> null
            }
        }
        # (3 unchanged blocks hidden)
    }

```

When applying the plan, is throwing:

```
commercetools_type.regular-customer: Modifying... [id=bfc50073-8989-4c96-8f4d-afb2cbfd93e0]
╷
│ Error: Request cancelled
│ 
│   with commercetools_type.regular-customer,
│   on types.tf line 14, in resource "commercetools_type" "regular-customer":
│   14: resource "commercetools_type" "regular-customer" {
│ 
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-commercetools_v1.0.1 plugin:

panic: runtime error: slice bounds out of range [:4] with capacity 3

goroutine 73 [running]:
github.com/labd/terraform-provider-commercetools/commercetools.resourceTypeFieldChangeActions({0xc000247d80, 0x4, 0x0?}, {0xc0003ab650, 0x3, 0xc000393680?})
	github.com/labd/terraform-provider-commercetools/commercetools/resource_type.go:473 +0xb4f
github.com/labd/terraform-provider-commercetools/commercetools.resourceTypeUpdate({0x1138e60, 0xc000649b60}, 0xc0000d4a80, {0xfdcc40?, 0xc00000c528})
	github.com/labd/terraform-provider-commercetools/commercetools/resource_type.go:359 +0x690
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc000381ea0, {0x1138df0, 0xc000454140}, 0x24?, {0xfdcc40, 0xc00000c528})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:375 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000381ea0, {0x1138df0, 0xc000454140}, 0xc00044dee0, 0xc0000d5100, {0xfdcc40, 0xc00000c528})
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:475 +0x53a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc0001234d0, {0x1138df0, 0xc000454140}, 0xc0000a5540)
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/grpc_provider.go:977 +0xd8a
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0001a1200, {0x1138e98?, 0xc000468cf0?}, 0xc0000a6460)
	github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/tf5server/server.go:603 +0x2ad
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xfac800?, 0xc0001a1200}, {0x1138e98, 0xc000468cf0}, 0xc000146[420](xxx)
	github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000237340, {0x113b7d8, 0xc00019ca80}, 0xc00016aa00, 0xc0003ab7d0, 0x17449e0, 0x0)
	google.golang.org/grpc@v1.32.0/server.go:1194 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc000237340, {0x113b7d8, 0xc00019ca80}, 0xc00016aa00, 0x0)
	google.golang.org/grpc@v1.32.0/server.go:1517 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc@v1.32.0/server.go:859 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.32.0/server.go:857 +0x28a

Error: The terraform-provider-commercetools_v1.0.1 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

Error: Terraform exited with code 1.
Error: Process completed with exit code 1.
```

Is there any workaround we can apply here?

Thanks

--- a/commercetools/resource_type.go
+++ b/commercetools/resource_type.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"log"
+	"math"
 	"reflect"
 	"time"
 
@@ -470,7 +471,7 @@ func resourceTypeFieldChangeActions(oldValues []interface{}, newValues []interfa
 		newNames[i] = v["name"].(string)
 	}
 
-	if checkAttributeOrder && !reflect.DeepEqual(oldNames, newNames[:len(oldNames)]) {
+	if checkAttributeOrder && !reflect.DeepEqual(oldNames, newNames[:int(math.Min(float64(len(newNames)), float64(len(oldNames))))]) {
 		log.Printf("[DEBUG] Field ordering: %s", newNames)
 
 		actions = append(

--- a/commercetools/resource_type.go
+++ b/commercetools/resource_type.go
@@ -4,7 +4,6 @@ import (
 	"context"
 	"fmt"
 	"log"
-	"math"
 	"reflect"
 	"time"
 
@@ -471,7 +470,7 @@ func resourceTypeFieldChangeActions(oldValues []interface{}, newValues []interfa
 		newNames[i] = v["name"].(string)
 	}
 
-	if checkAttributeOrder && !reflect.DeepEqual(oldNames, newNames[:int(math.Min(float64(len(newNames)), float64(len(oldNames))))]) {
+	if checkAttributeOrder && !reflect.DeepEqual(oldNames, newNames[:min(len(newNames), len(oldNames))]) {
 		log.Printf("[DEBUG] Field ordering: %s", newNames)
 
 		actions = append(

--- a/commercetools/resource_type_test.go
+++ b/commercetools/resource_type_test.go
@@ -176,7 +176,11 @@ func TestAccTypes_UpdateWithID(t *testing.T) {
 					resource.TestCheckResourceAttr(
 						"commercetools_type.acctest_type", "key", name),
 					resource.TestCheckResourceAttr(
-						"commercetools_type.acctest_type", "field.#", "10"),
+						"commercetools_type.acctest_type", "field.#", "12"),
+					resource.TestCheckResourceAttr(
+						"commercetools_type.acctest_type", "field.3.name", "icq_uin"),
+					resource.TestCheckResourceAttr(
+						"commercetools_type.acctest_type", "field.4.name", "testing"),
 					resource.TestCheckResourceAttr(
 						"commercetools_type.acctest_type", "field.1.name", "existing_enum"),
 					resource.TestCheckResourceAttr(
@@ -335,6 +339,26 @@ resource "commercetools_type" "%s" {
 					nl = "Skype"
 				}
 			}
+		}
+	}
+
+	field {
+		name = "icq_uin"
+		label = {
+			en = "UIN"
+		}
+		type {
+			name = "String"
+		}
+	}
+
+	field {
+		name = "testing"
+		label = {
+			en = "test"
+		}
+		type {
+			name = "String"
 		}
 	}
 

--- a/commercetools/resource_type_test.go
+++ b/commercetools/resource_type_test.go
@@ -162,6 +162,8 @@ func TestAccTypes_UpdateWithID(t *testing.T) {
 					resource.TestCheckResourceAttr(
 						"commercetools_type.acctest_type", "key", name),
 					resource.TestCheckResourceAttr(
+						"commercetools_type.acctest_type", "field.0.name", "skype_name"),
+					resource.TestCheckResourceAttr(
 						"commercetools_type.acctest_type", "field.1.name", "existing_enum"),
 					resource.TestCheckResourceAttr(
 						"commercetools_type.acctest_type", "field.1.type.0.element_type.0.values.%", "2"),
@@ -174,13 +176,13 @@ func TestAccTypes_UpdateWithID(t *testing.T) {
 					resource.TestCheckResourceAttr(
 						"commercetools_type.acctest_type", "key", name),
 					resource.TestCheckResourceAttr(
-						"commercetools_type.acctest_type", "field.#", "11"),
+						"commercetools_type.acctest_type", "field.#", "10"),
 					resource.TestCheckResourceAttr(
-						"commercetools_type.acctest_type", "field.2.name", "existing_enum"),
+						"commercetools_type.acctest_type", "field.1.name", "existing_enum"),
 					resource.TestCheckResourceAttr(
-						"commercetools_type.acctest_type", "field.2.type.0.element_type.0.values.%", "3"),
+						"commercetools_type.acctest_type", "field.1.type.0.element_type.0.values.%", "3"),
 					resource.TestCheckResourceAttr(
-						"commercetools_type.acctest_type", "field.2.type.0.element_type.0.values.evening", "Evening Changed"),
+						"commercetools_type.acctest_type", "field.1.type.0.element_type.0.values.evening", "Evening Changed"),
 				),
 			},
 		},
@@ -275,17 +277,6 @@ resource "commercetools_type" "%s" {
 	}
 
 	resource_type_ids = ["customer"]
-
-	field {
-		name = "skype_name"
-		label = {
-			en = "Skype name"
-			nl = "Skype naam"
-		}
-		type {
-			name = "String"
-		}
-	}
 
 	field {
 		name = "new_enum"

--- a/commercetools/utils.go
+++ b/commercetools/utils.go
@@ -406,3 +406,10 @@ func compareDateString(a, b string) bool {
 func diffSuppressDateString(k, old, new string, d *schema.ResourceData) bool {
 	return compareDateString(old, new)
 }
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}

